// redux-app-config/config.js
export const version = {
  default: {},
  persist: false,
  actions: {
    SET_VERSION: {
      reducer: 'SET',
    },
  },
};

const defaultUser = {
  settings: { unit: 'cm', stand: 110, sit: 80 },
  reminder: { sitMinute: 45, standMinute: 15 },
  user: '',
};

export const userInfo = {
  default: defaultUser,
  persist: true,
  actions: {
    UPDATE_USER_SETTINGS: {
      inputs: ['unit', 'stand', 'sit'],
      reducer: (state, { payload }) => ({
        ...state,
        settings: { ...state.settings, ...payload },
      }),
    },
    UPDATE_REMINDER: {
      inputs: ['sitMinute', 'standMinute'],
      reducer: (state, { payload }) => ({
        ...state,
        reminder: { ...state.reminder, ...payload },
      }),
    },
  },
};

export const bleStatus = {
  default: {
    status: 'BLE_OFF',
  },
  persist: false,
  actions: {
    UPDATE_BLE_STATUS: {
      inputs: ['status'],
      reducer: 'MERGE',
    },
  },
};

export const peskInfo = {
  default: {
    height: 0,
    error: null,
    locked: false,
  },
  persist: false,
  actions: {
    SET_HEIGHT: {
      inputs: ['height'],
      reducer: (state, { payload }) => ({
        ...state,
        height: payload.height,
        error: null,
      }),
    },
    SET_LOCK_STATUS: {
      inputs: ['locked'],
      reducer: (state, { payload }) => ({
        ...state,
        ...payload,
      }),
    },
    SET_PESK_ERROR: {
      reducer: 'MERGE',
    },
  },
};
